js/blocked_number_controller.js

```diff
@@ -0,0 +1,85 @@
+/* global , Whisper, storage, ConversationController */
+/* global textsecure: false */
+
+/* eslint-disable more/no-then */
+
+// eslint-disable-next-line func-names
+(function() {
+    'use strict';
+
+    window.Whisper = window.Whisper || {};
+
+    const blockedNumbers = new Whisper.BlockedNumberCollection();
+    window.getBlockedNumbers = () => blockedNumbers;
+
+    window.BlockedNumberController = {
+      getAll() {
+        try {
+          this.load();
+        } catch (e) {
+          console.warn(e);
+        }
+        return blockedNumbers;
+      },
+      reset() {
+        blockedNumbers.reset([]);
+      },
+      load() {
+        window.log.info('BlockedNumberController: starting initial fetch');
+
+        if (blockedNumbers.length) {
+          throw new Error('BlockedNumberController: Already loaded!');
+        }
+
+        if (!storage) {
+          throw new Error('BlockedNumberController: Could not load blocked numbers');
+        }
+
+        // Add the numbers to the collection
+        const numbers = storage.getBlockedNumbers();
+        blockedNumbers.add(
+          numbers.map(number => ({ number }))
+        );
+      },
+      block(number) {
+        const ourNumber = textsecure.storage.user.getNumber();
+
+        // Make sure we don't block ourselves
+        if (ourNumber === number) {
+          window.log.info('BlockedNumberController: Cannot block yourself!');
+          return null;
+        }
+
+        storage.addBlockedNumber(number);
+
+        // Make sure we don't add duplicates
+        const exists = blockedNumbers.getNumber(number);
+        if (exists)
+          return exists;
```

Comment:
I'm not sure what the controller's `block()` and `unblock()` should return, as the return value does not seem to be used?
It doesn't seem super consistent, unless I'm missing the pattern?

Reply:
I made them both return the `BlockNumber` model incase in the future we need to use it after we block/unblock.

Reply:
Maybe 'exists' should be existingUser or something along those lines, reads a bit strange because the `if (exists)` makes it feel like it's a bool?

Reply:
Ok, let's call YAGNI as I don't see how the BlockNumber model will provide any useful functionality in the future (it just holds a number) compared to the actual number as a string.
If we decide to add some functionality to it in the future, then we can change our block/unblock functions to return something useful.
For now, let's return bool if the number was successfully blocked/unblocked.
What do you think?